### PR TITLE
DOI citations

### PR DESCRIPTION
**Pull request by dchiquito**
Use the DOI for the citation URL if it is present.

### CODE REVIEW

File: dandiapi/api/tests/test_version.py [hunk 1]
```diff
@@ -104,13 +105,24 @@ def test_published_version_metadata_computed(published_version):
...
+        draft_version.metadata['citation']
+        == f'{name} ({year}). (Version {draft_version.version}) [Data set]. DANDI archive. {url}'  # noqa: E501
+    )
+
+
+@pytest.mark.django_db
+def test_version_metadata_citation_published(published_version):
+    name = published_version.metadata['name'].rstrip('.')
+    year = datetime.now().year
+    url = f'https://doi.org/{published_version.doi}'
```
Comment (satra) [inline dandiapi/api/tests/test_version.py:122]: are you actually calling the production server for this? or mocking it?

Comment (satra) [reply dandiapi/api/tests/test_version.py:122]: i.e is `published_version` a fixture that spoofs a doi?

**Review by satra — APPROVED** [review]
the code looks good to me.

just left a question about `published_version` fixture.

File: dandiapi/api/tests/test_version.py (hunk 1 above)
Comment (dchiquito) [reply dandiapi/api/tests/test_version.py:122]: The DOI credentials are not set up in the test environment. When publishes occur in testing, a DOI string is generated, but no request is made to canonize that as a real DOI.

File: dandiapi/api/tests/factories.py [hunk 2]
```diff
@@ -76,6 +77,19 @@ class PublishedVersionFactory(BaseVersionFactory):
     class Meta:
         model = Version
 
+    @classmethod
+    def _create(cls, *args, **kwargs):
+        version: Version = super()._create(*args, **kwargs)
+        version.doi = f'10.80507/dandi.{version.dandiset.identifier}/{version.version}'
```
Comment (waxlamp) [inline dandiapi/api/tests/factories.py:83]: Where does that hardcoded DOI prefix come from? Should it be parameterized somehow?

I realize this is a testing module, so maybe it doesn't matter here, but if this value corresponds to a production datacite instance or something like that, I wonder if it should be stored in and retrieved from an environment variable or something like that.

**Review by waxlamp — APPROVED** [review]

File: dandiapi/api/tests/factories.py (hunk 2 above)
Comment (dchiquito) [reply dandiapi/api/tests/factories.py:83]: https://github.com/dandi/dandi-api/issues/472

**Pull request merged**